File: app_backend/fastapi_app/routers/plan.py
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel, Field
from sqlmodel import Session

from ..auth import get_current_user
from ..db import get_session
from ..models import Plan, User, UserRead

router = APIRouter(prefix="/plan", tags=["Plan"])


class PlanResponse(BaseModel):
    """Response schema for returning the current plan."""
    plan: Plan = Field(..., description="The user's current subscription plan")


class UpdatePlanRequest(BaseModel):
    """Request schema for updating the user's plan."""
    plan: Plan = Field(..., description="New plan value")


# PUBLIC_INTERFACE
@router.get(
    "",
    summary="Get user plan",
    description="Returns the current subscription plan for the authenticated user.",
    response_model=PlanResponse,
    responses={401: {"description": "Unauthorized"}},
)
def get_plan(current_user: Annotated[User, Depends(get_current_user)]) -> PlanResponse:
    """Return the authenticated user's current plan."""
    return PlanResponse(plan=current_user.plan)


# PUBLIC_INTERFACE
@router.put(
    "",
    summary="Update user plan",
    description="Change the authenticated user's subscription plan.",
    response_model=UserRead,
    responses={400: {"description": "Validation error"}, 401: {"description": "Unauthorized"}},
)
def update_plan(
    payload: UpdatePlanRequest,
    current_user: Annotated[User, Depends(get_current_user)],
    session: Annotated[Session, Depends(get_session)],
) -> UserRead:
    """Update the user's plan and return the updated user profile."""
    if payload.plan not in {Plan.normal, Plan.premium, Plan.ultra}:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Invalid plan")

    current_user.plan = payload.plan
    session.add(current_user)
    session.commit()
    session.refresh(current_user)

    return UserRead(
        id=current_user.id,
        email=current_user.email,
        plan=current_user.plan,
        created_at=current_user.created_at,
        updated_at=current_user.updated_at,
    )

File: app_backend/src/routes/plan.js
import { Router } from "express";
import { body, validationResult } from "express-validator";
import { requireAuth } from "../middleware/auth.js";
import { getSequelize } from "../config/database.js";

const router = Router();

/**
 * @openapi
 * /api/plan:
 *   get:
 *     tags: [Plan]
 *     security: [{ bearerAuth: [] }]
 *     summary: Get user plan
 *     description: Returns the current subscription plan for the authenticated user.
 *     responses:
 *       200:
 *         description: Plan object
 *         content:
 *           application/json:
 *             schema:
 *               type: object
 *               properties:
 *                 plan:
 *                   type: string
 *                   enum: [normal, premium, ultra]
 *       401:
 *         description: Unauthorized
 */
router.get("/", requireAuth, async (req, res, next) => {
  try {
    const sequelize = getSequelize();
    const User = sequelize.models.User;
    const user = await User.findByPk(req.auth.uid);
    if (!user) {
      return res.status(404).json({ message: "User not found" });
    }
    res.json({ plan: user.plan });
  } catch (err) {
    next(err);
  }
});

/**
 * @openapi
 * /api/plan:
 *   put:
 *     tags: [Plan]
 *     security: [{ bearerAuth: [] }]
 *     summary: Update user plan
 *     description: Change the authenticated user's subscription plan.
 *     requestBody:
 *       required: true
 *       content:
 *         application/json:
 *           schema:
 *             type: object
 *             required: [plan]
 *             properties:
 *               plan:
 *                 type: string
 *                 enum: [normal, premium, ultra]
 *     responses:
 *       200:
 *         description: Updated user object
 *         content:
 *           application/json:
 *             schema:
 *               $ref: '#/components/schemas/User'
 *       400:
 *         description: Validation error
 *       401:
 *         description: Unauthorized
 */
router.put(
  "/",
  requireAuth,
  [body("plan").isIn(["normal", "premium", "ultra"]).withMessage("Invalid plan")],
  async (req, res, next) => {
    try {
      const errors = validationResult(req);
      if (!errors.isEmpty()) {
        return res.status(400).json({ message: "Validation error", details: errors.array() });
      }
      const { plan } = req.body;

      const sequelize = getSequelize();
      const User = sequelize.models.User;
      const user = await User.findByPk(req.auth.uid);
      if (!user) {
        return res.status(404).json({ message: "User not found" });
      }

      user.plan = plan;
      await user.save();

      res.json({
        id: user.id,
        email: user.email,
        plan: user.plan,
        createdAt: user.createdAt,
        updatedAt: user.updatedAt,
      });
    } catch (err) {
      next(err);
    }
  }
);

export default router;
